### PR TITLE
[OPIK-4222] [BE] Create dataset_evaluators table and API endpoints

### PR DESCRIPTION
**Pull request by JetoPistola**
## Details

This PR implements the `dataset_evaluators` table and API endpoints for the Evaluation Suite feature (OPIK-4109).

### Changes:
- **Database Migration**: Added Liquibase migration `000049_add_dataset_evaluators_table.sql` creating the `dataset_evaluators` table with:
  - Primary key on `id`
  - Composite index on `(workspace_id, dataset_id)` for efficient queries
  - `type` column as `ENUM('llm_judge', 'code_metric')` for evaluator type
  - `config` JSON column for flexible evaluator configuration

- **API Models**:
  - `DatasetEvaluator` - Main entity with nested `DatasetEvaluatorPage`, `DatasetEvaluatorBatchCreateRequest`, `DatasetEvaluatorCreate`
  - `EvaluatorType` - Java enum (`LLM_JUDGE`, `CODE_METRIC`) matching MySQL ENUM values

- **DAO Layer**: `DatasetEvaluatorDAO` with batch insert, find by dataset ID (paginated), find by IDs, and batch delete operations.

- **JDBI Type Handling**: `EvaluatorTypeArgumentFactory` and `EvaluatorTypeColumnMapper` in `infrastructure/db/` for JDBI enum serialization.

- **Service Layer**: `DatasetEvaluatorService` with workspace isolation and transaction management

- **REST Endpoints** at `/v1/private/datasets/{datasetId}/evaluators`:
  - `POST /batch` - Batch create evaluators
  - `GET` - Get evaluators (paginated)
  - `POST /delete-batch` - Batch delete evaluators

- **Validation**:
  - Jakarta `@ExecutionPolicyValidation` custom validator to enforce `passThreshold <= runsPerItem` on `ExecutionPolicy`
  - `@Max(100)` cap on both `runsPerItem` and `passThreshold` fields
  - Unit tests for `ExecutionPolicyValidator` covering valid, cross-field, min, and max boundary cases

- **Integration Tests**: Full test coverage for all endpoints including workspace isolation tests

### Recent Improvements:
- **Revision 6**: Standardized log message formatting with consistent quoting around numeric placeholders
- **Revision 7**: Enhanced batch request validation - added `@NotNull @Valid` on list elements to enforce null-item rejection and nested validation of `DatasetEvaluatorCreate` fields
- **Revision 8**: Added dataset existence validation in `createBatch` and `getByDatasetId` to prevent orphan records and ensure proper 404 responses for missing datasets
- **Revision 9**: Replaced manual test fixture construction with `PodamFactory.manufacturePojo()` + `toBuilder()` pattern for consistency with other tests; added `toBuilder = true` to `DatasetEvaluatorCreate`
- **Revision 10**: Replaced count+delete with single delete in `deleteBatch` — uses returned row count to validate all IDs existed, removing the separate `countByIdsAndDatasetId` DAO method
- **Revision 11**: Fixed workspace isolation test to expect 404 (consistent with Revision 8's dataset existence validation)
- **Revision 12**: Per scope update — renamed `metric_type` to `type` (now `ENUM('llm_judge', 'code_metric')`), renamed `metric_config` to `config`. Added `EvaluatorType` Java enum with JDBI argument factory and column mapper.
- **Revision 13**: Fixed inline fully-qualified imports (`java.util.stream.Collectors`, `java.util.stream.IntStream`) to proper imports. Added dataset existence validation to `deleteBatch` for consistency.
- **Revision 14**: Added 404 tests for non-existent dataset on all operations (create, get, delete) to verify parent resource validation.
- **Revision 15**: Added Jakarta `ExecutionPolicyValidation` / `ExecutionPolicyValidator` to enforce `passThreshold <= runsPerItem` on `ExecutionPolicy`.
- **Revision 16**: Added `@Max(10)` cap on both `runsPerItem` and `passThreshold` in `ExecutionPolicy`.
- **Revision 17**: Added parameterized unit tests for `ExecutionPolicyValidator` covering valid combos, cross-field violations, `@Min` boundary, and `@Max` boundary cases (16 tests).
- **Revision 18**: Increased `@Max` cap from 10 to 100 on both `runsPerItem` and `passThreshold` to support small datasets with more runs. Updated test boundaries accordingly.
- **Revision 19**: Renamed `DatasetEvaluatorBatchRequest` to `DatasetEvaluatorBatchCreateRequest` for clarity.
- **Revision 20**: Extracted `EvaluatorTypeArgumentFactory` and `EvaluatorTypeColumnMapper` from nested DAO classes to standalone files in `infrastructure/db/`, consistent with other type handlers (`AlertType`, `JsonNode`, etc.).

## Change checklist
- [x] User facing
- [ ] Documentation update

## Issues
- OPIK-4222

## Testing
- Added comprehensive integration tests in `DatasetEvaluatorsResourceTest.java`
- Tests cover:
  - Batch creation of evaluators
  - Pagination of results
  - Batch deletion
  - Workspace isolation (evaluators from one workspace not visible in another)
  - 404 responses for non-existent parent dataset (create, get, delete)
- Added unit tests in `ExecutionPolicyValidatorTest.java` for all validation constraints

## Documentation
N/A - API documentation is auto-generated from OpenAPI annotations

### CODE REVIEW

File: apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/DatasetEvaluatorsResource.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+@Path("/v1/private/datasets/{datasetId}/evaluators")
+@Produces(MediaType.APPLICATION_JSON)
+@Consumes(MediaType.APPLICATION_JSON)
+@Timed
+@Slf4j
+@RequiredArgsConstructor(onConstructor_ = @Inject)
+@Tag(name = "Dataset Evaluators", description = "Dataset Evaluator resources")
+public class DatasetEvaluatorsResource {
+
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/DatasetEvaluatorsResource.java:49]: Hey Baz

Validation IS being enforced at runtime through Dropwizard's configuration. Adding Spring's @Validated would be inconsistent with the codebase and wouldn't work without adding Spring Validation dependencies.

File: apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/DatasetEvaluatorsResource.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        String workspaceId = requestContext.get().getWorkspaceId();
+
+        log.info("Deleting {} dataset evaluators for dataset '{}' on workspaceId '{}'",
+                request.ids().size(), datasetId, workspaceId);
+
+        service.deleteBatch(new HashSet<>(request.ids()));
+
+        log.info("Deleted dataset evaluators for dataset '{}' on workspaceId '{}'",
+                datasetId, workspaceId);
+
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/DatasetEvaluatorsResource.java:118]: Hey Baz, we addressed this

File: apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+@Builder(toBuilder = true)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+public record DatasetEvaluator(
+        @JsonView( {
+                View.Public.class, View.Write.class}) UUID id,
+        @JsonView({View.Public.class, View.Write.class}) @NotNull UUID datasetId,
+        @JsonView({View.Public.class, View.Write.class}) @NotBlank @Size(max = 255) String name,
+        @JsonView({View.Public.class, View.Write.class}) @NotBlank @Size(max = 100) String metricType,
+        @JsonView({View.Public.class, View.Write.class}) @NotNull JsonNode metricConfig,
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java]: Hey Baz

Recommendation: Use String
Keep the metricType field as a String for these reasons:
Extensibility: The SDK has many built-in metrics (Hallucination, AnswerRelevance, ContextPrecision, Moderation, Factuality, GEval, etc.) and users can create custom metrics. Using an enum would require backend changes every time a new metric is added to the SDK.
SDK Independence: The backend shouldn't need to know all possible evaluator types - it's just storing configuration for the SDK to use later.
Existing Pattern: Similar to how AutomationRuleEvaluator stores complex configurations with a type field that can vary, DatasetEvaluator stores SDK metric configurations that the client controls.
Dynamic Nature: The Python SDK has 30+ metric classes (Hallucination, AnswerRelevance, ContextPrecision, ROUGE, BLEU, BERTScore, etc.) and supports user-defined custom metrics.

File: apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/DatasetEvaluatorsResource.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    @GET
+    @Operation(operationId = "getDatasetEvaluators", summary = "Get dataset evaluators", description = "Get paginated list of evaluators for a dataset", responses = {
+            @ApiResponse(responseCode = "200", description = "Dataset evaluators page", content = @Content(schema = @Schema(implementation = DatasetEvaluatorPage.class)))
+    })
+    @JsonView(DatasetEvaluator.View.Public.class)
+    public Response getEvaluators(
+            @PathParam("datasetId") UUID datasetId,
+            @QueryParam("page") @Min(1) @DefaultValue("1") int page,
+            @QueryParam("size") @Min(1) @DefaultValue("10") int size) {
+
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/DatasetEvaluatorsResource.java:86]: Hi Baz

Believe that this issue relates to other places in code too

For consistency, let's keep as is in the rest of the codebase

File: apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    public record DatasetEvaluatorPage(
+            @JsonView( {
+                    View.Public.class}) List<DatasetEvaluator> content,
+            @JsonView({View.Public.class}) int page,
+            @JsonView({View.Public.class}) int size,
+            @JsonView({View.Public.class}) long total) implements Page<DatasetEvaluator>{
+
+        public static DatasetEvaluatorPage empty(int page) {
+            return new DatasetEvaluatorPage(List.of(), page, 0, 0);
+        }
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java:54]: Hey Baz

List.of() returns a cached singleton of an empty list

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        return template.inTransaction(WRITE, handle -> {
+            var dao = handle.attach(DatasetEvaluatorDAO.class);
+            dao.batchInsert(evaluators, workspaceId);
+
+            Set<UUID> ids = evaluators.stream().map(DatasetEvaluator::id).collect(java.util.stream.Collectors.toSet());
+            List<DatasetEvaluator> created = dao.findByIds(workspaceId, ids);
+
+            log.info("Created '{}' dataset evaluators for dataset '{}' in workspace '{}'",
+                    created.size(), datasetId, workspaceId);
+            return created;
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java:84]: Hey Baz

Logging inside transactions is fine — it's harmless on rollback, unlike emails or cache updates

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        datasetService.findById(datasetId, workspaceId, Visibility.PRIVATE);
+
+        log.info("Creating '{}' dataset evaluators for dataset '{}' in workspace '{}'",
+                request.evaluators().size(), datasetId, workspaceId);
+
+        List<DatasetEvaluator> evaluators = request.evaluators().stream()
+                .map(create -> {
+                    UUID id = idGenerator.generateId();
+                    IdGenerator.validateVersion(id, "DatasetEvaluator");
+                    return DatasetEvaluator.builder()
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java:63]: Hey Baz

The validations are already in place:

createBatch: DatasetEvaluatorBatchRequest.evaluators has @NotNull @Size(min = 1, max = 1000), and the resource uses @NotNull @Valid on the request body — Jakarta validation rejects null/empty before the service is reached.

deleteBatch: The resource takes @NotNull @Valid BatchDelete request, which has its own validation on ids.

Service layer: Both parameters use Lombok's @NonNull, which throws NullPointerException if somehow null slips through.

A null collection can't reach the stream/dereference — it's caught at the API boundary by Jakarta validation. No change needed.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+@ImplementedBy(DatasetEvaluatorServiceImpl.class)
+public interface DatasetEvaluatorService {
+
+    List<DatasetEvaluator> createBatch(UUID datasetId, DatasetEvaluatorBatchRequest request);
+
+    DatasetEvaluatorPage getByDatasetId(UUID datasetId, int page, int size);
+
+    void deleteBatch(UUID datasetId, Set<UUID> ids);
+
+    void deleteByDatasetId(UUID datasetId);
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java:35]: Hi Baz

The service is intentionally synchronous — it uses the synchronous TransactionTemplate with JDBI for pure MySQL operations, and its callers are synchronous JAX-RS endpoints returning Response. Services in the codebase that use Mono/Flux do so because they interact with ClickHouse async DAOs and need to compose across reactive boundaries. Wrapping synchronous JDBI calls in Mono.fromCallable here would add complexity without benefit since the entire call chain is blocking. If a reactive caller is needed in the future, we can migrate then.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            if (matchingCount != ids.size()) {
+                throw new BadRequestException(
+                        "Some evaluator IDs do not belong to dataset '%s'. Requested: %d, found: %d"
+                                .formatted(datasetId, ids.size(), matchingCount));
+            }
+
+            int deleted = dao.deleteByIdsAndDatasetId(workspaceId, datasetId, ids);
+            log.info("Deleted '{}' dataset evaluators for dataset '{}' in workspace '{}'",
+                    deleted, datasetId, workspaceId);
+            return null;
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java]: Hey Baz

Wasn't this addressed?

[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java]: Hey Baz

Did you notice, this commit:

https://github.com/comet-ml/opik/pull/5076/commits/966f5358c0368e13479af96c54ecb2beb51b093a

[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java]: Think that you have some caching issue Baz, it was applied in a commit that was pushed to the code

File: apps/opik-backend/src/main/resources/liquibase/db-app-state/migrations/000049_add_dataset_evaluators_table.sql [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    name VARCHAR(255) NOT NULL,
+    type ENUM('llm_judge', 'code_metric') NOT NULL,
+    config JSON NOT NULL,
+    created_at TIMESTAMP(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6),
+    created_by VARCHAR(100) NOT NULL DEFAULT 'admin',
+    last_updated_at TIMESTAMP(6) NOT NULL DEFAULT CURRENT_TIMESTAMP(6) ON UPDATE CURRENT_TIMESTAMP(6),
+    last_updated_by VARCHAR(100) NOT NULL DEFAULT 'admin',
+    CONSTRAINT `dataset_evaluators_pk` PRIMARY KEY (id),
+    INDEX `dataset_evaluators_workspace_dataset_idx` (workspace_id, dataset_id)
+);
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/resources/liquibase/db-app-state/migrations/000049_add_dataset_evaluators_table.sql:18]: Hey Baz

Believe it's not needed here, because it's part of the table, and not a separate migration

File: apps/opik-backend/src/main/resources/liquibase/db-app-state/migrations/000049_add_dataset_evaluators_table.sql [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+--rollback DROP TABLE IF EXISTS dataset_evaluators;
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/resources/liquibase/db-app-state/migrations/000049_add_dataset_evaluators_table.sql:20]: Hey Baz

Every table creation migration uses --rollback DROP TABLE IF EXISTS. Our rollback is consistent with the rest of the codebase.

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java]: Hey Baz

Don't you see revision 10 in code?

[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorService.java]: Baz, I promise you we deleted this, I see in files changed, not countByIdsAndDataset under deleteBatch

File: apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Builder
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+    public record DatasetEvaluatorBatchRequest(
+            @NotNull @Size(min = 1, max = 1000) List<@NotNull @Valid DatasetEvaluatorCreate> evaluators) {
+    }
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java:62]: Hey Baz

It's consistent with the rest of the codebase. @Size(min = 1, max = 1000) is the standard pattern used across 23+ batch request types — SpanBatch, TraceBatch, ExperimentItemsBatch, BatchDelete, DatasetItemBatch, GuardrailBatch, etc.

Making this one configurable while all others are hardcoded would be an inconsistency. If there's ever a need to make batch sizes tunable, it should be done across the board, not just here.

File: apps/opik-backend/src/main/java/com/comet/opik/api/ExecutionPolicy.java [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+
+@Builder(toBuilder = true)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+public record ExecutionPolicy(
+        @Min(1) int runsPerItem,
+        @Min(1) int passThreshold) {
+
+    public static final ExecutionPolicy DEFAULT = new ExecutionPolicy(1, 1);
+}
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/api/ExecutionPolicy.java:20]: Hey Baz

We can add later, sent a message to product team on setting the Max cap

Did add, validation that passThreshold <= runsPerItem

File: apps/opik-backend/src/main/java/com/comet/opik/api/ExecutionPolicy.java [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+@Builder(toBuilder = true)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+@ExecutionPolicyValidation
+public record ExecutionPolicy(
+        @Min(1) @Max(10) int runsPerItem,
+        @Min(1) @Max(10) int passThreshold) {
+
+    public static final ExecutionPolicy DEFAULT = new ExecutionPolicy(1, 1);
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/api/ExecutionPolicy.java:19]: Hey Baz

DEFAULT is an API-level fallback for when users don't specify an execution policy — it's not a deployment configuration. Users who want different values set them explicitly per evaluator. Making this configurable via env/properties would conflate API defaults with deployment config, adding complexity for no real use case.

File: apps/opik-backend/src/main/java/com/comet/opik/api/ExecutionPolicy.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
+package com.comet.opik.api;
+
+import com.comet.opik.api.validation.ExecutionPolicyValidation;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.databind.PropertyNamingStrategies;
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/api/ExecutionPolicy.java:5]: Hey Baz

It's going to be part of the dataset / datasetItem tables, not the evaluator tables

File: apps/opik-backend/src/main/java/com/comet/opik/api/validation/ExecutionPolicyValidator.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+            ConstraintValidator<ExecutionPolicyValidation, ExecutionPolicy> {
+
+    @Override
+    public boolean isValid(ExecutionPolicy executionPolicy,
+            ConstraintValidatorContext constraintValidatorContext) {
+        if (executionPolicy == null) {
+            return true;
+        }
+        return executionPolicy.passThreshold() <= executionPolicy.runsPerItem();
+    }
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/main/java/com/comet/opik/api/validation/ExecutionPolicyValidator.java:18]: Hey Baz

Did you notice the new ExecutionPolicyValidatorTest.java?

File: apps/opik-backend/src/test/java/com/comet/opik/api/validation/ExecutionPolicyValidatorTest.java [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @ParameterizedTest
+    @MethodSource("invalidMinCases")
+    void validateWhenBelowMin(int runsPerItem, int passThreshold) {
+        var policy = new ExecutionPolicy(runsPerItem, passThreshold);
+
+        Set<ConstraintViolation<ExecutionPolicy>> violations = validator.validate(policy);
+
+        assertThat(violations).isNotEmpty();
```
[the remark this answers is not in the document]
Comment (JetoPistola) [reply apps/opik-backend/src/test/java/com/comet/opik/api/validation/ExecutionPolicyValidatorTest.java:55]: Hey Baz

I'd keep them separate. They test conceptually different constraints (@Min vs @Max), and separate methods give clearer failure messages — you immediately know which boundary was violated. Merging them into validateWhenOutOfBounds saves 3 lines but loses that clarity.

Not worth the change for a minor DRY concern in tests.

File: apps/opik-backend/src/main/resources/liquibase/db-app-state/migrations/000049_add_dataset_evaluators_table.sql [hunk 18]
```diff
@@ -0,0 +1,20 @@
+--liquibase formatted sql
+--changeset danieldimenshtein:000049_add_dataset_evaluators_table
+--comment: Create dataset_evaluators table for storing dataset-level SDK metric configurations
+
+CREATE TABLE IF NOT EXISTS dataset_evaluators (
+    id CHAR(36) NOT NULL,
+    workspace_id VARCHAR(150) NOT NULL,
+    dataset_id CHAR(36) NOT NULL,
+    name VARCHAR(255) NOT NULL,
```
Comment (thiagohora) [inline apps/opik-backend/src/main/resources/liquibase/db-app-state/migrations/000049_add_dataset_evaluators_table.sql:9]: I see that on the request, we only receive the name. I guess we should also allow users to define the ID, since it's created by the SDK. However,  if we want to retrieve it by name, we must adda  `UNIQUE KEY` by [(workspace_id, name)` or `(workspace_id, dataset_id, name)`

File: apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+        public static DatasetEvaluatorPage empty(int page) {
+            return new DatasetEvaluatorPage(List.of(), page, 0, 0);
+        }
+    }
+
+    @Builder
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+    public record DatasetEvaluatorBatchRequest(
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java]:
```suggestion
    public record DatasetEvaluatorBatchCreateRequest(
```

File: apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+            @NotNull @Size(min = 1, max = 1000) List<@NotNull @Valid DatasetEvaluatorCreate> evaluators) {
+    }
+
+    @Builder(toBuilder = true)
+    @JsonIgnoreProperties(ignoreUnknown = true)
+    @JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+    public record DatasetEvaluatorCreate(
+            @NotBlank @Size(max = 255) String name,
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java:68]: Should this be unique? If want to retrieve it by name we have to consider that

File: apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+import java.time.Instant;
+import java.util.List;
+import java.util.UUID;
+
+@Builder(toBuilder = true)
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
+public record DatasetEvaluator(
+        @JsonView( {
+                View.Public.class, View.Write.class}) UUID id,
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java:24]: If the client cannot provide it, it must be read-only. But I believe we should allow users(SDK) to define it. At least this has been the pattern so far.

File: apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+        @JsonView({View.Public.class}) @Schema(accessMode = Schema.AccessMode.READ_ONLY) Instant lastUpdatedAt,
+        @JsonView({View.Public.class}) @Schema(accessMode = Schema.AccessMode.READ_ONLY) String lastUpdatedBy){
+
+    public static class View {
+        public static class Public {
+        }
+
+        public static class Write {
+        }
+    }
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/api/DatasetEvaluator.java:40]: Do we need this sense? We have a different object to create.

File: apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/DatasetEvaluatorsResource.java [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+import java.util.UUID;
+
+@Path("/v1/private/datasets/{datasetId}/evaluators")
+@Produces(MediaType.APPLICATION_JSON)
+@Consumes(MediaType.APPLICATION_JSON)
+@Timed
+@Slf4j
+@RequiredArgsConstructor(onConstructor_ = @Inject)
+@Tag(name = "Dataset Evaluators", description = "Dataset Evaluator resources")
+public class DatasetEvaluatorsResource {
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/DatasetEvaluatorsResource.java:48]: Can evaluators be modified?

File: apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorDAO.java [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+        }
+    }
+
+    class EvaluatorTypeColumnMapper implements ColumnMapper<EvaluatorType> {
+        @Override
+        public EvaluatorType map(ResultSet r, int columnNumber, StatementContext ctx) throws SQLException {
+            String value = r.getString(columnNumber);
+            return value == null ? null : EvaluatorType.fromString(value);
+        }
+
```
Comment (thiagohora) [inline apps/opik-backend/src/main/java/com/comet/opik/domain/DatasetEvaluatorDAO.java]: Extract those methods from the DAO, so as to  avoid mixing it with DB infra details

**Comment by JetoPistola** [comment]
We decided to add 2 columns, evaluators & executionPolicy, to both dataset_versions & dataset_item_versions, so:

1. We won’t create new tables
2. We will build on top of the dataset versioning algorithm.

Hence, closing this PR, will use PR for executionPolicy code
